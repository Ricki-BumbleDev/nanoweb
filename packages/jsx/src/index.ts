///<reference path="./typings.d.ts" />
const rxUnescaped = new RegExp(/["'&<>]/);

export const render = async (element: any): Promise<string> => {
  const awaited = element instanceof Promise ? await element : element;
  if (awaited instanceof Template) {
    return '<!DOCTYPE html>' + (await resolve(awaited));
  }
  return element;
};

const resolve = async (elements: any): Promise<string> => {
  let result = '';
  if (Array.isArray(elements)) {
    for (const child of elements) {
      result += await resolve(child instanceof Promise ? await child : child);
    }
  } else if (elements || elements === 0) {
    return elements;
  }
  return result;
};

function escape(text: any): string {
  /* Much faster when there is no unescaped characters */
  if (typeof text !== 'string' || !rxUnescaped.test(text)) {
    return text;
  }

  let result = '';
  let escape = '';
  let start = 0;
  let i: number;
  for (i = 0; i < text.length; ++i) {
    switch (text.charCodeAt(i)) {
      case 34: // "
        escape = '&quot;';
        break;
      case 39: // '
        escape = '&#039;';
        break;
      case 38: // &
        escape = '&amp;';
        break;
      case 60: // <
        escape = '&lt;';
        break;
      case 62: // >
        escape = '&gt;';
        break;
      default:
        continue;
    }
    if (i > start) {
      result += text.slice(start, i);
    }
    result += escape;
    start = i + 1;
  }
  return result + text.slice(start, i);
}

export class Template extends Array {}

export type TemplateType = typeof Template;

export const unsafe = (value = '') => Template.from([value]);

const SELF_CLOSING = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
]);

export const jsx = (type: any, props: Record<string, any>): Template => {
  if (typeof type === 'function') {
    return type(props);
  } else if (typeof type === 'undefined') {
    if (props.children) {
      return Template.from(props.children);
    } else {
      return new Template();
    }
  }
  let propString = '';
  if (props) {
    for (const i in props) {
      if (props[i] != null) {
        propString += toString(i, props[i]);
      }
    }
  }
  if (props.children == null) {
    const template = new Template();
    if (SELF_CLOSING.has(type)) {
      template[0] = '<' + type + propString + '/>';
    } else {
      template[0] = '<' + type + propString + '></' + type + '>';
    }
    return template;
  }
  const children =
    Array.isArray(props?.children) && !(props.children instanceof Template) ? props.children : [props.children];
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (!(child instanceof Template)) {
      children[i] = escape(child);
    }
  }
  const template = new Template();
  const length = children.length;
  template[0] = '<' + type + propString + '>';
  for (let i = 0; i < length; i++) {
    template[i + 1] = children[i];
  }
  template[length + 1] = '</' + type + '>';
  return template;
};

export const jsxs = jsx;
export const Fragment = (props: Record<string, any>) => jsx(undefined, props);

const CAMEL_CASE_PATTERN = new RegExp('[a-z]+((d)|([A-Z0-9][a-z0-9]+))*([A-Z])?', 'g');

const toKebapCase = (value: string) => value.match(CAMEL_CASE_PATTERN)!.join('-');

const kebapCache: Record<string, string> = {};
const keyCache: Record<string, string> = {};

const toString = (key: string, value: any) => {
  if (value == null || key === 'children') {
    return '';
  } else if (key === 'class' && Array.isArray(value)) {
    value = value.flat().filter(Boolean).join(' ');
  } else if (key === 'style') {
    value = Object.entries(value)
      .map(entry => [kebapCache[entry[0]] || (kebapCache[entry[0]] = toKebapCase(entry[0])), entry[1]].join(':'))
      .join(';')
      .toLowerCase();
  } else if (typeof value === 'object') {
    value = JSON.stringify(value);
  } else if (typeof value === 'boolean') {
    return ` ${keyCache[key] || (keyCache[key] = escape(key.toLowerCase()))}`;
  }
  return ` ${keyCache[key] || (keyCache[key] = escape(key.toLowerCase()))}="${escape(value)}"`;
};
