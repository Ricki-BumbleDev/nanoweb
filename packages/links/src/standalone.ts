import 'instant.page';
import links from './links';

window.addEventListener('DOMContentLoaded', () => {
  links({ defaultLoadingAnimation: true });
});
