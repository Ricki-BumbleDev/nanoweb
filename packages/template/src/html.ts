const repeatedWhitespace = /\s{2,}/g;
const whitespace = /\r\n|\n|\r|(\s{1,}(?=<))/g;

const htmlEscape = (value: any) =>
  typeof value !== 'string' || (value as any) instanceof UnsafeHtml
    ? value
    : value
        .replace(/&/g, '&amp;')
        .replace(/>/g, '&gt;')
        .replace(/</g, '&lt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#39;')
        .replace(/`/g, '&#96;');

const cache = new Map();

type AtomicValues = false | string | number | undefined | null | Template | UnsafeHtml;

export type Child = AtomicValues | Promise<AtomicValues> | Promise<AtomicValues[]>;

export type TemplateElement = Child | Child[];

export class Template extends Array<Child> {
  public webComponents: string[];

  constructor(webComponents?: string[]) {
    super();
    this.webComponents = webComponents || [];
  }
}

class UnsafeHtml extends String {}

export const unsafeHtml = (value = '') => new UnsafeHtml(value);

const WEB_COMPONENT_PATTERN = /(?:<| is=")([a-zA-Z0-9]+-[a-zA-Z0-9-]+)/g;

const resolve = (element: TemplateElement): Template | Promise<Template> => {
  const template = new Template();
  if (element instanceof Template) {
    return element;
  } else if (Array.isArray(element)) {
    // TODO consider using Template.from here
    template.push(...element.map(resolve));
  } else if (element instanceof Promise) {
    template.push((element as any).then(resolve));
  } else {
    template.push(htmlEscape(element));
  }
  return template;
};

export const html = (literalSections: TemplateStringsArray, ...substs: TemplateElement[]) => {
  let [raw, webComponents] = cache.get(literalSections) || [undefined, undefined];
  if (raw === undefined) {
    raw = literalSections.raw.map((item: string) => item.replace(whitespace, '').replace(repeatedWhitespace, ' '));
    webComponents = Array.from(
      new Set(Array.from(raw.join().matchAll(WEB_COMPONENT_PATTERN)).map((hit: any) => hit[1])),
    );
    cache.set(literalSections, [raw, webComponents]);
  }
  const result = new Template(webComponents);
  for (let i = 0; i < substs.length; i++) {
    const subst = substs[i];
    result.push(raw[i]);
    if (subst instanceof Template) {
      result.push(...subst.flat());
      result.webComponents = [...result.webComponents, ...subst.webComponents];
    } else if (Array.isArray(subst)) {
      result.push(...subst.flatMap(htmlEscape));
      // TODO Check if needed
      subst.forEach(item => {
        if (item instanceof Template) {
          result.webComponents.push(...item.webComponents);
        }
      });
    } else if (subst instanceof Promise) {
      result.push(resolve(subst));
    } else {
      result.push(htmlEscape(subst));
    }
  }
  result.push(raw[raw.length - 1]);
  return result;
};
