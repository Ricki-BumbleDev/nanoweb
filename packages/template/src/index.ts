export * from './html';
export * from './render-to-string';
export * from './cache';
