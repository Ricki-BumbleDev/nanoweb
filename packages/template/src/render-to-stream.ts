import { PassThrough, Readable } from 'stream';
import { Template } from './html';

const resolve = async (stream: PassThrough, component: Template | Promise<Template>, webComponents: string[]) => {
  const list = await component;
  webComponents.push(...list.webComponents);
  for (const item of list) {
    if (item != null && item !== false) {
      const resolvedItem = await item;
      if (resolvedItem instanceof Template) {
        await resolve(stream, resolvedItem, webComponents);
      } else {
        stream.write(String(resolvedItem));
      }
    }
  }
};

export interface StreamRenderOptions {
  transformResult?: (text: Readable, webComponents: string[]) => Readable;
}

export const renderToStream = (component: Template | Promise<Template>, options?: StreamRenderOptions): Readable => {
  const sink = new PassThrough({});
  const webComponents: string[] = [];
  setImmediate(async () => {
    await resolve(sink, component, webComponents);
    sink.end();
  });
  if (options?.transformResult) {
    return options.transformResult(sink, [...new Set(webComponents)]);
  }
  return sink;
};
