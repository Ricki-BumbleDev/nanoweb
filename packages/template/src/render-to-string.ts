import { Template } from './html';

export const resolve = async (component: Template | Promise<Template>, webComponents: string[]) => {
  const list = await component;
  webComponents.push(...list.webComponents);
  let buffer = '';
  for (const item of list) {
    if (item !== undefined) {
      const resolvedItem = await item;
      buffer += resolvedItem instanceof Template ? await resolve(resolvedItem, webComponents) : resolvedItem;
    }
  }
  return buffer;
};

export interface RenderOptions {
  transformResult?: (text: string, webComponents: string[]) => string;
}

export const renderToString = async (component: Template | Promise<Template>, options?: RenderOptions) => {
  const webComponents: string[] = [];
  const buffer = await resolve(component, webComponents);
  if (options?.transformResult) {
    return options.transformResult(buffer, [...new Set(webComponents)]);
  }
  return buffer;
};
