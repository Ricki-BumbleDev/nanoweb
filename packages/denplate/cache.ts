import { Cache } from 'https://deno.land/x/local_cache@1.0/mod.ts';
import { Template } from './html.ts';
import { resolve } from './render-to-string.ts';

type CacheKeyFunction<TFn> = TFn extends (...a: infer A) => any ? (...a: A) => false | string : never;

type CacheResult = Template | Promise<Template>;

interface Options<T> {
  ttl?: number;
  cacheKey?: CacheKeyFunction<T>;
}

/**
 * Caches component responses in an optimized format
 */
export const cache = <T extends (...args: any[]) => CacheResult>(
  func: T,
  options: Options<T> = {},
): ((...funcArgs: Parameters<T>) => CacheResult) => {
  const { ttl, cacheKey } = options;
  const componentCache = new Cache<string, CacheResult>(ttl);
  return (...args: Parameters<T>): CacheResult => {
    const key = cacheKey?.(...args) || 'static';
    if (componentCache.has(key)) {
      return componentCache.get(key)!;
    } else {
      const result = func(...args);
      setImmediate(async () => {
        const webComponents: string[] = [];
        const compiledString = await resolve(result, webComponents);
        const compiledTemplate = new Template(webComponents);
        compiledTemplate[0] = compiledString;
        componentCache.set(key, compiledTemplate);
      });
      return result;
    }
  };
};
