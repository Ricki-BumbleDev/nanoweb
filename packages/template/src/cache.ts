import NodeCache from 'node-cache';
import { Template } from './html';
import { resolve } from './render-to-string';

type CacheKeyFunction<TFn> = TFn extends (...a: infer A) => any ? (...a: A) => false | string : never;

type CacheResult = Template | Promise<Template>;

interface Options<T> extends Omit<NodeCache.Options, 'useClones'> {
  cacheKey?: CacheKeyFunction<T>;
}

/**
 * Caches component responses in an optimized format
 */
export default function cache<T extends (...args: any[]) => CacheResult>(
  func: T,
  options: Options<T> = {},
): (...funcArgs: Parameters<T>) => CacheResult {
  const { cacheKey, ...nodeCacheOptions } = options;
  const componentCache = new NodeCache({ ...nodeCacheOptions, useClones: false });
  return (...args: Parameters<T>): CacheResult => {
    const key = cacheKey?.(...args) || 'static';
    if (componentCache.has(key)) {
      return componentCache.get<CacheResult>(key)!;
    } else {
      const result = func(...args);
      setImmediate(async () => {
        const webComponents: string[] = [];
        const compiledString = await resolve(result, webComponents);
        const compiledTemplate = new Template(webComponents);
        compiledTemplate[0] = compiledString;
        componentCache.set(key, compiledTemplate);
      });
      return result;
    }
  };
}
