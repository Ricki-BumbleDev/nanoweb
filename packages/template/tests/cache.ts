import cache from '../src/cache';
import { html, Template, renderToString, RenderOptions } from '../src/index';

const nextTick = () => new Promise(resolve => setImmediate(resolve));

const matchSnapshot = async (getComponent: () => Template | Promise<Template>, options?: RenderOptions) =>
  expect(await renderToString(getComponent(), options)).toMatchSnapshot();

describe('Cache templates', () => {
  test('without options', async () => {
    let counter = 0;
    const Component = () => html`Counter: ${++counter}`;
    const cachedComponent = cache(Component);
    await matchSnapshot(cachedComponent);
    await nextTick();
    await matchSnapshot(cachedComponent);
  });

  test('with TTL option', async () => {
    let counter = 0;
    const Component = () => html`Counter: ${++counter}`;
    const cachedComponent = cache(Component, { stdTTL: 0.1 });
    await matchSnapshot(cachedComponent);
    await nextTick();
    await matchSnapshot(cachedComponent);
    await new Promise(resolve => setTimeout(resolve, 120));
    await matchSnapshot(cachedComponent);
  });

  test('with cache key generator', async () => {
    let counter = 0;
    const Component = (id: string) => html`ID: ${id}, Counter: ${++counter}`;
    const cachedComponent = cache(Component, { cacheKey: (id: string) => id });
    await matchSnapshot(() => cachedComponent('A'));
    await nextTick();
    await matchSnapshot(() => cachedComponent('B'));
    await nextTick();
    await matchSnapshot(() => cachedComponent('A'));
  });
});
