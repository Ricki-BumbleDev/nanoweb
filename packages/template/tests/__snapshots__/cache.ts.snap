// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cache templates with TTL option 1`] = `"Counter: 1"`;

exports[`Cache templates with TTL option 2`] = `"Counter: 1"`;

exports[`Cache templates with TTL option 3`] = `"Counter: 2"`;

exports[`Cache templates with cache key generator 1`] = `"ID: A, Counter: 1"`;

exports[`Cache templates with cache key generator 2`] = `"ID: B, Counter: 2"`;

exports[`Cache templates with cache key generator 3`] = `"ID: A, Counter: 1"`;

exports[`Cache templates without options 1`] = `"Counter: 1"`;

exports[`Cache templates without options 2`] = `"Counter: 1"`;
